File: 8. BucketSort.cpp
/*
8. Bucket Sort

Implement Bucket Sort for floating-point numbers in the range (0, 1).

Display the content of each bucket before and after sorting.
*/

#include <bits/stdc++.h>
using namespace std;

// Function to perform Bucket Sort
void bucketSort(vector<float> &arr) {
    int n = arr.size();
    vector<vector<float>> buckets(n);

    // Distribute elements into buckets
    for(float num : arr) {
        int index = n * num; // bucket index
        buckets[index].push_back(num);
    }

    // Display buckets before sorting
    cout << "Buckets before sorting:\n";
    for(int i = 0; i < n; i++) {
        cout << "Bucket " << i << ": ";
        for(float num : buckets[i]) cout << num << " ";
        cout << endl;
    }

    // Sort individual buckets
    for(int i = 0; i < n; i++)
        sort(buckets[i].begin(), buckets[i].end());

    // Display buckets after sorting
    cout << "\nBuckets after sorting:\n";
    for(int i = 0; i < n; i++) {
        cout << "Bucket " << i << ": ";
        for(float num : buckets[i]) cout << num << " ";
        cout << endl;
    }

    // Merge buckets into original array
    int idx = 0;
    for(int i = 0; i < n; i++)
        for(float num : buckets[i])
            arr[idx++] = num;
}

int main() {
    vector<float> arr = {0.78, 0.17, 0.39, 0.26, 0.72, 0.94, 0.21, 0.12, 0.23, 0.68};
    
    cout << "Original array: ";
    for(float num : arr) cout << num << " ";
    cout << endl;

    bucketSort(arr);

    cout << "\nSorted array: ";
    for(float num : arr) cout << num << " ";
    cout << endl;

    return 0;
}

/*
Type: Non-comparison-based, Stable

Time Complexity:

Best Case: O(n + k) → When elements are uniformly distributed across buckets
Average Case: O(n + k) → n = number of elements, k = number of buckets
Worst Case: O(n²) → When all elements fall into a single bucket

Space Complexity:
Auxiliary Space: O(n + k) → Extra space for buckets

*/

File: 8. BucketSort.py
def bucket_sort(arr):
    """
    Sorts a list of floats in the range [0, 1) using Bucket Sort.

    Type: Non-comparison-based, Stable

    Time Complexity:
    - Best Case: O(n + k) -> When elements are uniformly distributed.
    - Average Case: O(n + k) -> n = elements, k = buckets.
    - Worst Case: O(n^2) -> When all elements fall into one bucket.

    Space Complexity:
    - Auxiliary Space: O(n + k) -> For the buckets.
    """
    n = len(arr)
    # 1. Create n empty buckets
    buckets = [[] for _ in range(n)]

    # 2. Distribute array elements into different buckets
    for num in arr:
        # Calculate bucket index. Multiplying by n maps the float
        # to an index in the range [0, n-1].
        index = int(n * num)
        buckets[index].append(num)

    # Display buckets before sorting them individually
    print("Buckets before sorting:")
    for i, bucket in enumerate(buckets):
        # Using a generator expression with ' '.join for cleaner printing
        bucket_content = ' '.join(map(str, bucket))
        print(f"Bucket {i}: {bucket_content}")

    # 3. Sort individual buckets
    for bucket in buckets:
        bucket.sort() # Python's Timsort is used here

    # Display buckets after sorting
    print("\nBuckets after sorting:")
    for i, bucket in enumerate(buckets):
        bucket_content = ' '.join(map(str, bucket))
        print(f"Bucket {i}: {bucket_content}")
    
    # 4. Concatenate all buckets back into arr
    idx = 0
    for bucket in buckets:
        for num in bucket:
            arr[idx] = num
            idx += 1

def main():
    """Main function to test the bucket_sort."""
    arr = [0.78, 0.17, 0.39, 0.26, 0.72, 0.94, 0.21, 0.12, 0.23, 0.68]
    
    print(f"Original array: {' '.join(map(str, arr))}\n")
    
    bucket_sort(arr)
    
    print(f"\nSorted array: {' '.join(map(str, arr))}")

if __name__ == "__main__":
    main()
    """
    Sorts a list of floats in the range [0, 1) using Bucket Sort.

    Type: Non-comparison-based, Stable

    Time Complexity:
    - Best Case: O(n + k) -> When elements are uniformly distributed.
    - Average Case: O(n + k) -> n = elements, k = buckets.
    - Worst Case: O(n^2) -> When all elements fall into one bucket.

    Space Complexity:
    - Auxiliary Space: O(n + k) -> For the buckets.
    """